File: src/main/java/com/braydenoneal/blang/parser/expression/builtin/GetItemsBuiltin.java
package com.braydenoneal.blang.parser.expression.builtin;

import com.braydenoneal.blang.parser.Program;
import com.braydenoneal.blang.parser.expression.Expression;
import com.braydenoneal.blang.parser.expression.value.ItemValue;
import com.braydenoneal.blang.parser.expression.value.ListValue;
import com.braydenoneal.blang.parser.expression.value.Value;
import net.minecraft.block.entity.LockableContainerBlockEntity;
import net.minecraft.item.Items;

import java.util.ArrayList;
import java.util.List;

public record GetItemsBuiltin(List<Expression> arguments) implements Expression {
    @Override
    public Value<?> evaluate(Program program) {
        List<LockableContainerBlockEntity> containers = program.context().entity().getConnectedContainers();
        List<Value<?>> items = new ArrayList<>();

        for (LockableContainerBlockEntity container : containers) {
            container.iterator().forEachRemaining(stack -> {
                if (!stack.isOf(Items.AIR)) {
                    items.add(new ItemValue(stack.getItem()));
                }
            });
        }

        return new ListValue(items);
    }
}

File: src/main/kotlin/com/braydenoneal/blang/parser/expression/builtin/GetItemsBuiltin.kt
package com.braydenoneal.blang.parser.expression.builtin

import com.braydenoneal.blang.parser.Program
import com.braydenoneal.blang.parser.expression.Arguments
import com.braydenoneal.blang.parser.expression.Expression
import com.braydenoneal.blang.parser.expression.ExpressionType
import com.braydenoneal.blang.parser.expression.ExpressionTypes
import com.braydenoneal.blang.parser.expression.value.ItemValue
import com.braydenoneal.blang.parser.expression.value.ListValue
import com.braydenoneal.blang.parser.expression.value.Value
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.item.Items


data class GetItemsBuiltin(val arguments: Arguments) : Expression {
    override fun evaluate(program: Program): Value<*> {
        val items: MutableList<Value<*>> = ArrayList()

        for (container in program.context().entity!!.getConnectedContainers()) {
            container.iterator().forEachRemaining { stack ->
                if (!stack.isOf(Items.AIR) && !items.contains(ItemValue(stack.item))) {
                    items.add(ItemValue(stack.item))
                }
            }
        }

        return ListValue(items)
    }

    override val type: ExpressionType<*> get() = ExpressionTypes.GET_ITEMS_BUILTIN

    companion object {
        val CODEC: MapCodec<GetItemsBuiltin> = RecordCodecBuilder.mapCodec {
            it.group(
                Arguments.CODEC.fieldOf("arguments").forGetter(GetItemsBuiltin::arguments)
            ).apply(it, ::GetItemsBuiltin)
        }
    }
}
